research/src/pages/popup/popup.proposal.mdx

```diff
@@ -0,0 +1,418 @@
+depending on whether or not the `popup` element is currently shown:
+
+```html
+<button id="menuButton" popup="menuPopup">Menu</button>
+<popup id="menuPopup" role="menu" anchor="menuButton">
+  <!-- Markup for menuitems goes here -->
+</popup>
+```
+
+The `popup` attribute is supported on a subset of
+[interactive elements](https://html.spec.whatwg.org/multipage/interactive-elements.html):
+
+- `button`
+- `input` in the `button` state
+- `input` in the `email`, `number`, `search`, `tel`, `text`, or `url` states
+
+#### Showing/hiding a `popup` element via the `popup` attribute
+
+When an element with the `popup` attribute specified receives a user interaction that should show
+the related `popup` element, this element shall be known as the `popup` element's _invoker_.
+
+<p class="note">
+  Multiple elements may have a popup attribute that refers to the same popup element. A given popup
+  element only has 0 or 1 invokers at a given time. For example, if the popup element was shown
+  programmatically, the invoker is null. Otherwise, the element that the user actually interacted
+  with to show the popup is its sole invoker.
+</p>
+
+If the _invoker_ is a `button` element or an `input` element in the `button` state, and the
+_invoker_ is activated; OR the _invoker_ is an input in the `email`, `number`, `search`, `tel`,
+`text`, or `url` states and focus is set to the _invoker_, the user agent must run the following
+steps:
+
+1. Let _subject_ be the element with the ID referenced by the `popup` attribute specified on the
+   _invoker_. If there is no such element, or the _subject_ is not a `popup` element,
+   then return.
+2. If the `open` IDL attribute on the _subject_ is `true`, run
+   [hiding a `popup` element steps](#hiding-current-popups).
+3. Otherwise, run [showing a `popup` element steps](#showing-popup-steps).
+
+<p class="question">
+  Should we have a way to suppress these steps upon focus, so that the author can enable the popup
+  to be shown upon text entry (or some other logic) instead?
+</p>
+
+<h3 id="showing-popup-steps">
+  Showing a <code>popup</code> element steps
+</h3>
+
+Let _candidate subject_ be the new candidate `popup` element to be shown.
+If applicable, let _invoker_ be the element whose `popup` attribute specifies the ID of
+the _candidate subject_, and which has been activated by the user.
+If applicable, let the _anchor element_ be the element whose ID is referenced by the `anchor`
+specified on the _candidate subject_.
+
+1. Run the [hiding currently-shown `popup` elements steps](#hiding-current-popups).
+2. Add the _subject_ to _subject_'s
+   [node document](https://dom.spec.whatwg.org/#concept-node-document)'s [popup stack](#popup-stack).
+3. Set the `open` IDL attribute on the _subject_ to `true`.
+4. If the `popup` element was shown as a result of a user interaction with an `invoker`, internally
+   set the _subject_'s _invoker_ to a pointer to the `invoker` element.
```

Comment:
Can you help me better understand your feedback on this one? Per your feedback on line 216, we'd already be passing in the _invoker_ parameter, we just need to make sure that relationship is stored elsewhere so that it can be referenced later by hide steps...would it be preferable to provide that directive where "showing a popup element steps" are called, vs as a step in the algorithm itself?

Reply:
Sure. The idea is that instead of "If the `popup` element was shown as a result of a user interaction with an `invoker`", this should be something like "If _dueToUserInteraction_ is true". Then, all call sites of the "showing a popup element steps" algorithm either pass true or false for that parameter.

Reply:
Thanks Domenic! Another way perhaps to solve this is to say "If _invoker_ is not null, internally set the _candidate subject_'s _invoker_ to a pointer to the `invoker` element." In this case, per the changes in ccd9b75036cb67e177ad11b2bf72b0b9c1ed0afd, we've specified that these steps take an optional _invoker_ param, and all relevant call sites pass that parameter. Thoughts on that vs. just adding another paremeter?